tests/testthat/test-crunch-data-frame.R

```diff
@@ -0,0 +1,184 @@
+context("CrunchDataFrame Methods")
+
+
+with_mock_crunch({
+    ds <- loadDataset("test ds")
+    
+    test_that("can manipulate the row order of a crunchDataFrame", {
+        ds_df <- as.data.frame(ds)
+        gndr <- as.vector(ds$gender)
+        expect_equal(nrow(ds_df), 25)
+        # both reording and subsetting the dataset
+        new_order <- c(4,3,1,2)
+        attr(ds_df, "order") <- new_order
+        expect_equal(ds_df$gender, gndr[new_order])
+        expect_equal(nrow(ds_df), 4)
+        ds_df2 <- as.data.frame(ds, row.order = new_order)
+        expect_equal(ds_df2$gender, gndr[new_order])
+        expect_equal(nrow(ds_df2), 4)
+    })
+    
+    test_that("Extract methods work with CrunchDataFrames like they do with data.frames", {
+        ds_df <- as.data.frame(ds)
+        true_df <- as.data.frame(ds, force = TRUE)
+        # single column/row extraction
+        expect_equivalent(ds_df[1,],true_df[1,])
+        expect_equivalent(ds_df[,1],true_df[,1])
+        
+        # multiple columns/rows
+        expect_equivalent(ds_df[c(1,2),],true_df[c(1,2),])
+        expect_equivalent(ds_df[,c(1,2)],true_df[,c(1,2)])
+        
+        # single column extraction with characters
+        expect_equivalent(ds_df[,"location"],true_df[,"location"])
+        
+        # multiple columns with characters
+        expect_equivalent(ds_df[,c("mymrset","textVar")],true_df[,c("mymrset","textVar")])
+        
+        # columns/rows with logicals
+        expect_equivalent(ds_df[,c(TRUE, FALSE, TRUE, FALSE, TRUE, FALSE, TRUE)],
+                          true_df[,c(TRUE, FALSE, TRUE, FALSE, TRUE, FALSE, TRUE)])
+        expect_equivalent(ds_df[c(TRUE, rep(FALSE, 23), TRUE),],
+                          true_df[c(TRUE, rep(FALSE, 23), TRUE),])
+        
+        # both rows and columns
+        expect_equivalent(ds_df[c(1,3,5),"location"],true_df[c(1,3,5),"location"])
+        
+        
+        # $ methods
+        expect_equivalent(ds_df$starttime,true_df$starttime)
+        
+        # [[ methods
+        expect_equivalent(ds_df[[1]],true_df[[1]])
+        expect_equivalent(ds_df[["catarray"]],true_df[["catarray"]])
+        expect_equivalent(ds_df[[2]],true_df[[2]])
+        expect_equivalent(ds_df[["birthyr"]],true_df[["birthyr"]])
+    })
+    
+    test_that("Extract methods input checking", {
+        ds_df <- as.data.frame(ds)
+        
+        expect_error(ds_df["foo",],
+                     "row subsetting must be done with either numeric or a logical")
```

Comment:
(Acknowledging, of course, that the first example is pointless, and the second one only works because the data.frame has row names set, which is not a feature I care to implement for crdfs right now.)

Reply:
We can't currently subset with `CrunchLogicalExpr` we could add that as a functionality, although that depends on the code that makes `as.vector` and `which` work with `CrunchLogicalExpr` which is already done in the Conditional Transforms branch.

As for the `data.frame`s response, that is because all data.frames have rownames (by default it's characters of the row indices (eg `c("1", "2", "3")`), and because `"foo"` isn't there, we get a row of `NA`s). 

We can mimic this, but we'll need to do add a bit more here, I would vote for deferring both of these until after we merge the conditional transforms branch since the code is in there that will allow `CrunchLogicalExpr` to be used already and then we can decide if we want to mimic `data.frame`s behavior with default rownames / ability to add rownames, etc.